### DISCUSSION TITLE
Help !!

### DISCUSSION THREAD

**Discussion by sardor0320** [community/community#147332] [New to GitHub] [answered]
we were writing a project with our colleagues. but a stranger changed the codes in our repository and wrote a commit, but he has no such right. how can this be

**Comment by peme969** [comment, marked as answer]
You can rollback the commits. If you want, change the repo to a private one via the settings and then invite your colleagues personally.
That way, no one else cant **view** it only your colleagues can.
Use git revert to undo the commit while maintaining history.
 If you lack branch protection rules, anyone with push access can make direct commits. On GitHub: Go to Settings > Branches and check if Branch protection rules are configured for key branches like main or master.
Enable branch protection rules to:
   - Require pull requests for changes. 
   - Restrict who can push to the branch.
   - Enable approvals for code changes.

**Comment by sardor0320** [comment]
Can I push the changes he made from the commit history?

**Reply by peme969** [reply]
Yes, you can remove or revert the changes made by the unauthorized commit.
To do so, heres a step-by-step tutorial:
First, identify the commit hash of the unauthorized changes:
`git log`
Locate the commit hash (e.g., abc1234) of the unwanted commit.
### Option 1: Git Revert (Safe & Recommended)
This creates a new commit that undoes the changes introduced by the unwanted commit while preserving history.
`git revert <commit-hash>`
Example:
`git revert abc1234`
A new commit will be created that undoes the changes.
This method is safe and doesn't rewrite history, which is ideal for shared repositories.
### Option 2: Git Reset (Destructive, Use with Caution)
If you want to completely remove the commit from history and rewrite the branch:
`git reset --hard <commit-hash-before-unwanted-commit>`

Example:
`git reset --hard abc1233`
Then, force-push the changes to the remote repository:

`git push origin <branch-name> --force`
⚠ Warning: Force-pushing can rewrite history, which can be dangerous if others are working on the branch.


Check who has push access to your repository:
Go to Settings > Collaborators and teams (on GitHub).
Remove unauthorized users.

**Reply by ghostinhershell** [reply]
Hi @sardor0320 👋🏾,It looks like @peme969's response was the answer to your question here. Can you please mark their response as the answer? When you mark a question as an answer, GitHub will highlight the comment and replies to the comment to help visitors quickly find the answer.
![image](https://github.com/user-attachments/assets/b89e7161-c2d1-47e2-a058-242df633e6b2)
